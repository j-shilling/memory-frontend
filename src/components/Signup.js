import React from "react";
import { Button, Form, Input } from "semantic-ui-react";

const EMAIL_REGEX = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;

class Signup extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      name: "",
      email: "",
      password: "",
      confirmPassword: ""
    };
  }

  handleInputChange = e => {
    this.setState({ [e.target.name]: e.target.value });
  };

  emailErrors = () => {
    if (this.state.email.length === 0) {
      return "Required";
    } else if (!EMAIL_REGEX.test(this.state.email)) {
      return "Invalid email";
    }
    return null;
  };

  passwordErrors = () => {
    if (this.state.password.length === 0) {
      return "Required";
    }

    return null;
  };

  confirmPasswordErrors = () => {
    if (this.state.confirmPassword.length === 0) {
      return "Required";
    } else if (this.state.password !== this.state.confirmPassword) {
      return "Does not match";
    }

    return null;
  };

  render() {
    return (
      <Form>
        <Form.Input
          name="name"
          placeholder="Name"
          value={this.state.name}
          onChange={this.handleInputChange}
        />
        <Form.Input
          name="email"
          placeholder="email"
          value={this.state.email}
          onChange={this.handleInputChange}
          error={this.emailErrors()}
        />
        <Form.Input
          name="password"
          type="password"
          placeholder="Password"
          value={this.state.password}
          onChange={this.handleInputChange}
          error={this.passwordErrors()}
        />
        <Form.Input
          name="confirmPassword"
          type="password"
          placeholder="Confirm Password"
          value={this.state.confirmPassword}
          onChange={this.handleInputChange}
          error={this.confirmPasswordErrors()}
        />
        <Button type="submit">Submit</Button>
      </Form>
    );
  }
}

export default Signup;
