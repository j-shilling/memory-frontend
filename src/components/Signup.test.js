import React from "react";
import { shallow, mount } from "enzyme";

import Signup from "./Signup";

describe("<Signup />", () => {
  const mockData = {
    name: "New Name",
    email: "new.name@email.com",
    password: "p@ssw0rd",
    confirmPassword: "p@ssw0rd"
  };

  const getChangeEvent = (name, value = "") => {
    return { target: { name, value } };
  };

  const wrapper = shallow(<Signup />);
  const form = () => wrapper.find("Form").first();
  const message = () => wrapper.find("Form Message").first();
  const submit = () => wrapper.find("Form Button").first();

  const getInput = field => wrapper.find(`FormInput[name="${field}"]`).first();

  const buildInputGetter = field => {
    return () => getInput(field);
  };

  const resetWrapper = () => {
    const inputs = wrapper.find("Form FormInput");
    for (let i = 0; i < inputs.length; i++) {
      const input = inputs.at(i);
      input.simulate("change", getChangeEvent(input.prop("name")));
    }
  };

  const getInputSelector = field => `Form FormInput[name="${field}"]`;

  describe("User sees inputs", () => {
    resetWrapper();
    it("User sees a form element", () => {
      expect(wrapper).toContainExactlyOneMatchingElement("Form");
    });
    const testField = field => {
      expect(wrapper).toContainExactlyOneMatchingElement(
        getInputSelector(field)
      );
    };
    for (let field in mockData) {
      it(`User sees a ${field} input`, () => {
        testField(field);
      });
    }
  });
  describe("User can enter values", () => {
    resetWrapper();
    const testField = field => {
      const input = buildInputGetter(field);
      input().simulate("change", getChangeEvent(field, mockData[field]));
      expect(input().prop("value")).toEqual(mockData[field]);
    };

    for (let field in mockData) {
      it(`User can change ${field} input`, () => {
        testField(field);
      });
    }
  });
  describe("User can see which fields contain errors", () => {
    resetWrapper();
    describe("Blank fields have no errors", () => {
      const testField = field => {
        const input = buildInputGetter(field);
        input().simulate("change", getChangeEvent(field));
        expect(input().prop("error")).toBeFalsy();
      };
      for (let field in mockData) {
        it(`${field} has no errors when blank`, () => {
          testField(field);
        });
      }
    });
    describe("User must enter valid data", () => {
      const testField = (field, data, str) => {
        const input = buildInputGetter(field);
        input().simulate("change", getChangeEvent(field, data));

        expect(input().prop("error")).toBeTruthy();
        expect(form().prop("error")).toBeTruthy();
        expect(message().prop("list")).toContain(str);

        input().simulate("change", getChangeEvent(field));
      };

      describe("Checks the email field", () => {
        it("Checks that email address is valid", () => {
          testField(
            "email",
            "invalid_address",
            "Please enter a valid email address"
          );
        });
      });
      describe("Checks the password field", () => {
        it("Checks that password has at least 8 characters", () => {
          testField(
            "password",
            "a1%",
            "Password should be at least 8 characters"
          );
        });
        it("Checks that password has at least 1 letter", () => {
          testField(
            "password",
            "12345678%",
            "Password should contain at least one letter"
          );
        });
        it("Checks that password has at least 1 number", () => {
          testField(
            "password",
            "abcdefg%",
            "Password should contain at least one number"
          );
        });
        it("Checks that password has at least 1 special character", () => {
          testField(
            "password",
            "abcdefg1",
            "Password should contain at least one of the following special characters: (~!@#$%^&*_-+=`|(){}[]:;\"'<>,.?/)"
          );
        });
        it("Checks that password does not contain any spaces", () => {
          testField(
            "password",
            "abcdefg1% ",
            "Passwords should not include any whitespace characters (spaces, tabs, etc.)"
          );
        });
        it("Checks that password does not contain any tabs", () => {
          testField(
            "password",
            "abcdefg1%\t",
            "Passwords should not include any whitespace characters (spaces, tabs, etc.)"
          );
        });

        const checkSpecialCharacter = char => {
          const input = buildInputGetter("password");

          input().simulate(
            "change",
            getChangeEvent("password", `12345678a${char}`)
          );
          expect(input().prop("error")).toBeFalsy();
          input().simulate("change", getChangeEvent("password"));
        };

        for (const c of "~!@#$%^&*_-+=`|(){}[]:;\"'<>,.?/") {
          it(`Counts '${c}' as a special character`, () => {
            checkSpecialCharacter("%");
          });
        }
      });
      describe("Checks the confirm password field", () => {
        it("Checks that the confirm password field matches the password field", () => {
          testField("confirmPassword", "1234567a%", "Passwords do not match");
        });
      });
    });
  });
  describe("User sees a submit button", () => {
    resetWrapper();
    it("Submit button is enabled when all fields are filled in with valid data", () => {
      for (let field in mockData) {
        getInput(field).simulate(
          "change",
          getChangeEvent(field, mockData[field])
        );
      }
      expect(submit().prop("disabled")).toBeFalsy();
    });

    for (let target in mockData) {
      it(`Submit button is disabled if ${target} is blank`, () => {
        for (let field in mockData) {
          getInput(field).simulate(
            "change",
            getChangeEvent(field, field === target ? "" : mockData[field])
          );
        }

        expect(submit().prop("disabled")).toBeTruthy();
      });
    }

    const mockInvalidData = [
      { name: "email", value: "new.name" },
      {
        name: "password",
        value: "asdf",
        otherFields: [{ name: "confirmPassword", value: "asdf" }]
      },
      {
        name: "confirmPassword",
        value: mockData.password
          .split("")
          .reverse()
          .join(""),
        otherFields: [{ name: "password", value: mockData.password }]
      }
    ];

    for (let data of mockInvalidData) {
      it(`Submit button is disabled if ${data.name} contains errors`, () => {
        for (let field in mockData) {
          let value = "";
          if (field === data.name) {
            value = data.value;
          } else {
            const otherField = data.otherFields
              ? data.otherFields.find(info => field === info.name)
              : false;
            value = otherField ? otherField.value : mockData[field];
          }
          getInput(field).simulate("change", getChangeEvent(field, value));
        }
        expect(submit().prop("disabled")).toBeTruthy();
      });
    }
  });
});
