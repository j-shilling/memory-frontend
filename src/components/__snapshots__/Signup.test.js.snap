// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Signup /> Renders correctly 1`] = `
<Card
  centered={true}
>
  <CardContent>
    <CardContent>
      <CardHeader
        as="h1"
        textAlign="center"
      >
        Create an Account
      </CardHeader>
    </CardContent>
    <Form
      as="form"
      error={false}
    >
      <Divider />
      <FormInput
        as={[Function]}
        control={[Function]}
        error={false}
        label="Display Name"
        name="name"
        onChange={[Function]}
        placeholder="Name"
        required={true}
        value=""
      />
      <FormInput
        as={[Function]}
        control={[Function]}
        error={false}
        label="Email Address"
        name="email"
        onChange={[Function]}
        placeholder="Email"
        required={true}
        value=""
      />
      <FormInput
        as={[Function]}
        control={[Function]}
        error={false}
        label="Enter Password"
        name="password"
        onChange={[Function]}
        placeholder="Password"
        required={true}
        type="password"
        value=""
      />
      <FormInput
        as={[Function]}
        control={[Function]}
        error={false}
        label="Confirm Password"
        name="confirmPassword"
        onChange={[Function]}
        placeholder="Confirm Password"
        required={true}
        type="password"
        value=""
      />
      <Message
        error={true}
        header="Fix the follow problems to submit:"
        list={Array []}
      />
      <Button
        as="button"
        disabled={true}
        fluid={true}
        type="submit"
      >
        Submit
      </Button>
      <Divider />
      <FormGroup>
        <span>
          Already have an account?Â 
        </span>
        <Link
          to="/login"
        >
          Login
        </Link>
      </FormGroup>
    </Form>
  </CardContent>
</Card>
`;
